### DISCUSSION TITLE
Default value for Variable Reference in Scripts

### DISCUSSION THREAD

**Discussion by aascedu** [netbox-community/netbox#17989] [Help Wanted!] [answered]
Hello,

I am working on a script to create VLAN Groups, VLANs, Prefixes.

I would like to be able to use a default value for the VLAN Prefix.
Let's say 192.168.101.0/24 for our Vlan Data :

```python
data_prefix = IPNetworkVar(
    label="Prefix Vlan Data",
    description="Type de prefix pour Vlan Data",
    default="192.168.101.0/24",
    required=False,
)
```
As I see in Netbox documentation : https://netbox.readthedocs.io/en/stable/customization/custom-scripts/#default-options:
![image](https://github.com/user-attachments/assets/4148bbab-3f0e-4daf-8b28-032529f13cf6)

However it never works... I can't get the default value. Even if the field would be a StringVar...
I tested it here :
```python
test=StringVar(
    default="testttt",
    required=False,
)
```
and tested calling this : `self.log_info(f"Test :{data['test']}")`

result :

`Test :`

expected result :

`Test :testttt`

**Comment by candlerb** [comment]
Netbox version?

This script works for me under Netbox 4.1.6:

```
from extras.scripts import Script, IPNetworkVar

class ShowPrefix(Script):
    class Meta:
        name = "Show Prefix"
        commit_default = False

    prefix = IPNetworkVar(
        description="Prefix",
        required=False,
        default="1.2.3.4/24",
    )

    def run(self, data, commit):
        return "prefix: %s" % repr(data["prefix"])
```

Result:

![image](https://github.com/user-attachments/assets/89f408c6-0a32-42c3-bc61-8fc4fe0ef3e2)

and after correcting the (invalid) prefix so that the script can run:

![image](https://github.com/user-attachments/assets/2a0a9d20-892d-4971-bee6-578a2e360b89)

**Comment by aascedu** [comment]
Hi, i’m running Netbox 4.1.4

**Reply by candlerb** [reply]
And if you run the script I showed?

**Reply by aascedu** [reply]
Hello, I ran the script you gave me in my netbox instance.
I understood a problem :
With the exact same script you gave me, when I run it, I don't see the form, it goes straight to the execution of the script and it prints this :
![image](https://github.com/user-attachments/assets/0c92258d-77a2-4e57-bee6-ebd6aec0beac)
However when I click on `Show Prefix` in the top left of the picture right here :
![image](https://github.com/user-attachments/assets/f93b19f1-9b0c-402e-b16d-9ce2cdc85d95)

I can see the form like so :
![image](https://github.com/user-attachments/assets/82cef235-7d2b-4a83-a08d-3b38e54f0d3c)

Now I can run the script with the default value (I had to change it to a valid network address) :
![image](https://github.com/user-attachments/assets/d3c9e720-3ea2-499b-826c-d7a5e5e8edff)

So I tested again, adding a required variable so that I would see the form : 
```python
from extras.scripts import Script, IPNetworkVar, StringVar

class ShowPrefix(Script):
    class Meta:
        name = "Show Prefix"
        commit_default = False

    prefix = IPNetworkVar(
        description="Prefix",
        required=False,
        default="1.2.3.0/24",
    )

    test = StringVar()

    def run(self, data, commit):
        self.log_info(data["test"])
        return "prefix: %s" % repr(data["prefix"])
```

And same behavior, if I run the script the prefix field is not loaded with the default value : 
![image](https://github.com/user-attachments/assets/78eceaca-7ab2-4f4b-8615-651df60d0204)
But if I click on the source tab in the top left corner of the picture and go back to the script form page, the default value pops up : 
![image](https://github.com/user-attachments/assets/849cb3b9-d9d4-4e7c-ad85-0f93cf20eebe)

I read the releases infos from all the versions of netbox 4-x-x and it doesn't seem that my version is the reason, I even updated my netbox to 4.1.6 and it did not fix anything.

Did your default value showed at first when you ran your script ?

**Reply by candlerb** [reply, marked as answer]
> Did your default value showed at first when you ran your script ?

Yes. But it depends what you mean by "ran your script".

I ran it by going to Customizations > Scripts, clicking on the script name in the left-hand column:

![image](https://github.com/user-attachments/assets/75cacded-5666-4b6c-911c-c1511cb3f64a)

This gives the data entry form, then I clicked Run Script at the bottom of the form.

Maybe you clicked the "Run Script" button on the right (or "Run Again" as it shows now); if so, it doesn't prompt for anything.

![image](https://github.com/user-attachments/assets/b44e6cb7-c196-494e-afc6-d01bab7ae794)

![image](https://github.com/user-attachments/assets/963ea189-25c3-4baf-b299-f687a35e3ef1)

If you were to say that's confusing and/or not useful, I wouldn't argue. Perhaps it should only be shown for scripts that don't declare any parameters.

**Reply by aascedu** [reply]
Hello, I feel dumb and at the same time Netbox made me dumb by showing me a big button with "Run again" on it.
Thanks so much 👍
